### ISSUE TITLE
Format expected value types when displaying in views

### ISSUE THREAD

**Issue by reefdog** [PhilanthropyDataCommons/front-end#52]
While demoing for @kfogel, he immediately observed how much of the data looked wrong (paraphrasing) due to lack of contextual formatting.

I recognize the responsibility of this formatting will be shared with the PDC itself, which may need to supplement field data with information about its type, e.g. what currency a given budget field uses. So: the API should probably be providing a schema.

For tracking and discussion, though, here are specific types Karl called out as worthy of consistent formatting in the template:

- **Currencies**: Displayed with currency unit and localized format, e.g. `$32,123.93`
- **Dates/Times**: Displayed consistently as ISO 8601 and to the significance implied by the specific field. (I.e., a "Review Date" field shouldn't display a time; a "Year" field shouldn't display months or days. No field should display an ambiguous format like `10-01-01`.)
- **Phone numbers**: Displayed with country code and formatted to the most common country-specific format.
- **Long text** (e.g., mission statements): Displayed with Markdown formatting, perhaps? At least line breaks respected.
- **URLs/Emails**: Clickable links.
- **Files:** Downloadable links.
   - The actual "files" we display on many proposals currently weren't added to the PDC as files; those are literally filename strings that were submitted to those fields. Naturally we can't/won't link those since there's nothing to link them to.
- **EINs/Tax IDs**: Formatted to context-specific format (e.g., `##-#######` for US EINs)
- **Multiple values** (e.g., proposal 75's "Review Labels" or "Review Assigned To"): Displayed with some sort of delimiter / as a list.

After whatever discussion we have about this, I expect we'll split off some sub-tasks.

**Comment by jasonaowen** [comment]
Yes, a lot of this depends on the API. Here are some related issues:

- https://github.com/PhilanthropyDataCommons/service/issues/285
- https://github.com/PhilanthropyDataCommons/service/issues/283
- https://github.com/PhilanthropyDataCommons/service/issues/282
- https://github.com/PhilanthropyDataCommons/service/issues/216

Others I haven't opened issues for yet:
- **Currencies** will need either currency-specific fields (like `proposal_budget_usd`) or we need to depend on the API (and thus on the input process) to format appropriately - as we're doing now, with the inevitable attendant data quality issues
- **Long text** is a good note, and solely the responsibility of the front end
- **URLs/Emails** is another good note; we should also open a data validation issue on the API, because I think we'll soon have data quality issues if we don't already
- **Files**: we've [discussed this a bit on Zulip](https://chat.opentechstrategies.com/#narrow/stream/66-PDC/topic/Data/near/159383); this is purely a data quality issue right now
- **Multiple values**: the schema has support for this, but it must be used at data import time; the specific value mentioned is newline-delimited but stuffed into a single value

**Comment by jasonaowen** [comment]
On Friday, @kfogel made the executive decision that we should assume all currencies are in USD, and format appropriately.

We also discussed moving forward on the front-end before the back-end is ready; this would involve hardcoding certain base fields as containing values of an expected type, and then processing values as if they were that type, and showing some kind of error highlighting if that processing failed (such as showing the raw, unparsable value in red). The type hardcoding would be thrown away once the API was sending us that data, and the error highlighting would be thrown away once the API is validating data (as all processing should then succeed), but the processing itself should carry over.

I've spent some time thinking about how to do this in practice, and the approach I'm contemplating right now (cc @reefdog to sanity check this proposal) is to override the `dataType` key of canonical fields as part of the API response handling, pass that through to the proposal list and proposal detail, and then do the actual parsing/formatting inside `<ProposalTable>` or so. That is, the `<ProposalDetail>` page would pass through a value like `{ shortCode: 'organization_phone', ..., dataType: 'phone-number', value: '+14565551212' }`, and later `<ProposalTable>` would key on `dataType: 'phone-number'` and attempt to show it. I still need to iterate on how this'll work in more detail, but if this sounds bonkers @reefdog please speak up!

**Comment by reefdog** [comment]
@jasonaowen To paraphrase that to make sure I understand: until the API is providing data types, we're going to decorate API responses with them, based on a hard-coded mapping that we maintain. (Which mapping we'll throw away when the API provides it.) Any view which uses the API data will then be able to make display/formatting decisions based on the data type. Right?

If so, that sounds sensible to me. The mapping and decorating will be disposable, but everything else will still be usable.

**Comment by slifty** [comment]
Per #988 we aren't going to be supporting this functionality.
